docs/docs/01-ibc/03-apps/00-ibcv2apps.md

```diff
@@ -0,0 +1,307 @@
+
+```go
+// IBCModule implements the application interface given the keeper.
+// The implementation of the IBCModule interface could for example be in a file called ibc_module.go,
+// but ultimately file structure is up to the developer
+type IBCModule struct {
+  keeper keeper.Keeper
+}
+```
+
+Additionally, in the `module.go` file, add the following line:
+
+```go
+var (
+  _ module.AppModule      = AppModule{}
+  _ module.AppModuleBasic = AppModuleBasic{}
+  // Add this line
+  _ porttypes.IBCModule   = IBCModule{}
+)
+```
+
+### Packet callbacks
+
+IBC expects modules to implement callbacks for handling the packet lifecycle, as defined in the `IBCModule` interface.
+
+With IBC v2, modules are not directly connected. Instead a pair of clients are connected and register the counterparty clientID. Packets are routed to the relevant application module by the portID registered in the Router. Relayers send packets between the routers/packet handlers on each chain. 
+
+![IBC packet flow diagram](./images/packet_flow_v2.png)
+
+Briefly, a successful packet flow works as follows:
+
+1. A user sends a message to the IBC packet handler
+2. The IBC packet handler validates the message, creates the packet and stores the commitment and returns the packet sequence number. The [`Payload`](https://github.com/cosmos/ibc-go/blob/fe25b216359fab71b3228461b05dbcdb1a554158/proto/ibc/core/channel/v2/packet.proto#L26-L38), which contains application specific data, is routed to the relevant application.
+3. If the counterparty writes an acknowledgement of the packet then the sending chain will process the acknowledgement.
+4. If the packet is not successfully received before the timeout, then the sending chain processes the packet's timeout.
+
+#### Sending packets
+
+[`MsgSendPacket`](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/v2/types/tx.pb.go#L69-L75) is sent by a user to the [channel v2 message server](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/v2/keeper/msg_server.go), which calls `ChannelKeeperV2.SendPacket`. This validates the message, creates the packet, stores the commitment and returns the packet sequence number. The application must specify its own payload which is used by the application and sent with `MsgSendPacket`.
+
+An application developer needs to implement the custom logic the application executes when a packet is sent.  
+
+```go
+
+// OnSendPacket logic
+func (im *IBCModule) OnSendPacket(
+  ctx sdk.Context, 
+  sourceChannel string, 
+  destinationChannel string, 
+  sequence uint64, 
+  payload channeltypesv2.Payload, 
+  signer sdk.AccAddress) error {
+
+// implement any validation
+
+// implement payload decoding and validation
+
+// call the relevant keeper method for state changes as a result of application logic
+
+// emit events or telemetry data
+
+return nil
+}
+
+```
+
+#### Receiving packets
+
+To handle receiving packets, the module must implement the `OnRecvPacket` callback. This gets
```

Comment:
Me be worth calling out that it is important for an application to validate and confirm support for both version and encoding of the receiving packet